xls/dslx/stdlib/round.x

```diff
@@ -0,0 +1,459 @@
+                    //   - If we kept at least one integer bit, check for sign change.
+                    if rounding_all_bits { carry } else { sign_changed }
+                },
+                // Negative argument with +1 adjustment never overflows (-1 -> 0 etc.).
+                (Sign::Negative, true) => false,
+                (Sign::Negative, false) => {
+                    // Negative argument with no adjustment.
+                    // When every bit is rounded away the result is 0 (RTZ/RNE/RTP)
+                    // or -1 (RTN/RNA). The latter is not representable because no integer bits
+                    // remain.
+                    if rounding_all_bits &&
+                    (rounding_mode == RoundingMode::RTN || rounding_mode == RoundingMode::RNA) {
+                        true
+                    } else {
+                        false
+                    }
+                },
+            }
+        };
+
+        let rounded_u = if sign_changed {
+            // handles cases like:
+            // argument is two's complement, positive, and rounding away all bits
+            // RNE(0.9375) = 1 -> overflow
+            // round(RoundingMode::RNE, 4, NonNegative, s5:0b0_1111))
+            // without this correction, result would be s5:0b1_0000, i.e. -1
+            uN[N]:0
+        } else {
+            rounded_u
+        };
+        (overflow, rounded_u as xN[S][N])
+    }
+}
+
+// Rounds an unsigned integer:
+//  - rounds a runtime-specified number (`num_bits_rounded`) of least significant bits,
+//  - returns the full-width rounded result with the least significant `num_bits_rounded` bits
+//    zeroed.
+// Returns (overflow, rounded result).
+pub fn round_u<N: u32, W_NBR: u32 = {std::clog2(N + u32:1)}>
+    (rounding_mode: RoundingMode, num_bits_rounded: uN[W_NBR], unrounded: uN[N]) -> (u1, uN[N]) {
+    round(rounding_mode, num_bits_rounded, Sign::NonNegative, unrounded)
+}
+
+// Rounds an unsigned integer:
+// - rounds a compile-time-constant (`num_bits_rounded`) number of least significant bits,
+// - returns only the most significant bits (i.e., the rounded result), discarding the rounded-off
+//   bits.
+// Returns (overflow, rounded result).
+pub fn round_trunc_u
+    <NumBitsRounded: u32, N: u32, R: u32 = {N - NumBitsRounded},
+     W_NBR: u32 = {std::clog2(N + u32:1)}>
+    (rounding_mode: RoundingMode, unrounded: uN[N]) -> (u1, uN[R]) {
+    const_assert!(NumBitsRounded <= N);
+    let (overflow, rounded) = round_u(rounding_mode, NumBitsRounded as uN[W_NBR], unrounded);
+    let (rounded_msbs, _) = std::split_msbs<R>(rounded);
+    (overflow, rounded_msbs)
+}
+
+// Rounds an unsigned integer:
+// - such that after rounding it is `AtMost` (a compile-time constant) bits wide,
+// - returns only the most significant bits (i.e., the rounded result), discarding the rounded-off
+//   bits.
+// Returns (overflow, rounded result).
+pub fn round_trunc_to_u
```

Comment:
do we expect `R`, `NumBitsRounded` and `W_NBR` to be overridable by the user?

If no, maybe we should just move them as const in the function body? (since the return type doesn't depend on them)